RFS/src/main/java/org/opensearch/migrations/bulkload/common/OpenSearchClient.java

```diff
@@ -99,17 +129,76 @@ private Version versionFromResponse(HttpResponse resp) {
 
             var distroNode = versionNode.get("distribution");
             if (distroNode != null && distroNode.asText().equalsIgnoreCase("opensearch")) {
-                versionBuilder.flavor(Flavor.OPENSEARCH);
+                versionBuilder.flavor(getLikelyOpenSearchFlavor());
             } else {
                 versionBuilder.flavor(Flavor.ELASTICSEARCH);
             }
-            return versionBuilder.build();
+            return Mono.just(versionBuilder.build());
         } catch (Exception e) {
             log.error("Unable to parse version from response", e);
-            throw new OperationFailed("Unable to parse version from response: " + e.getMessage(), resp);
+            return Mono.error(new OperationFailed("Unable to parse version from response: " + e.getMessage(), resp));
         }
     }
 
+    Mono<Boolean> checkCompatibilityModeFromResponse(HttpResponse resp) {
+        if (resp.statusCode != 200) {
+            return Mono.error(new OperationFailed("Unexpected status code " + resp.statusCode, resp));
+        }
+        try {
+            var body = Optional.of(objectMapper.readTree(resp.body));
+            var persistentlyInCompatibilityMode = inCompatibilityMode(body.map(n -> n.get("persistent")));
+            var transientlyInCompatibilityMode = inCompatibilityMode(body.map(n -> n.get("transient")));
+            return Mono.just(persistentlyInCompatibilityMode || transientlyInCompatibilityMode);
+        } catch (Exception e) {
+            log.error("Unable to determine if the cluster is in compatibility mode", e);
+            return Mono.error(new OperationFailed("Unable to determine if the cluster is in compatibility mode from response: " + e.getMessage(), resp));
+        }
+    }
+    
+    private boolean inCompatibilityMode(Optional<JsonNode> node) {
+        return node.filter(n -> !n.isNull())
+            .map(n -> n.get("compatibility"))
+            .filter(n -> !n.isNull())
+            .map(n -> n.get("override_main_response_version"))
+            .filter(n -> !n.isNull())
+            .map(n -> n.asBoolean())
+            .orElse(false);
+    }
+
+    private Mono<Version> getVersionFromPlugins(HttpResponse resp) {
+        if (resp.statusCode != 200) {
+            return Mono.error(new OperationFailed("Unexpected status code " + resp.statusCode, resp));
+        }
+        try {
+            var body = (ArrayNode) objectMapper.readTree(resp.body);
+            var elemIterator = body.elements();
+            while (elemIterator.hasNext()) {
+                var current = elemIterator.next();
+                var isRepositoryS3 = Optional.ofNullable(current.get("component"))
+                    .filter(n -> !n.isNull())
+                    .map(n -> n.asText())
+                    .map(text -> "repository-s3".equals(text))
+                    .orElse(false);
+                if (isRepositoryS3) {
```

Comment:
What if they don't have the S3 Repo plugin installed?

Reply:
Interesting - a few follow-ups, then:
* Are we thinking this API will only be called against AOS or AOSS clusters as opposed to self-managed?
* Because the plugin version can be anything, where are we gaining confidence that it will continue to truly report the server version?  Is this best-effort convention or something more durable?

Reply:
Compatibility mode is a feature in the open-source, see [#847](https://github.com/opensearch-project/OpenSearch/pull/847), it can be enabled on self-managed clusters or amazon managed OpenSearch cluster.  

There are two reason that I trust repository-s3 plugin version value:
1) Repository-s3 plugin is one that is not modified for the Amazon deployment, its deployed as is (yay!).  Many other plugins are modified for special functionality in the AWS ecosystem (hence `unspecified` and `x.x.x.x` versions).
2) Repository-s3 plugin is built for bundling in the OpenSearch distributions where its version number will match OpenSearch's version number.

Maybe we tweak this over time after better APIs are added, but this seems like the best way to infer this information sans bug reports.

Reply:
Actually, nix this thread, I've switched to using the more reliable _cat/nodes api that also surfaces this information.